### PR TITLE
Improved the complexity of dequeue O(n) to O(1)

### PR DESCRIPTION
**Pull request by fahimfaisaal**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Removed the `view` method & added the `peekLast` and `toArray` method
- [x] Upgraded some documentation
- [x] Added some more test cases
- [x] Resolved - #942 

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
- [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: Data-Structures/Queue/Queue.js [hunk 1]
```diff
@@ -1,56 +1,121 @@
...
+* This implementation following the linked list structure.
 */
 
-// Functions: enqueue, dequeue, peek, view, length, empty
 class Queue {
-  // constructor
-  constructor () {
-    // This is the array representation of the queue
-    this.queue = []
+  #size
```
Comment (appgurueu) [inline Data-Structures/Queue/Queue.js:10]: Odd variable name; I suggest removing the #.

File: Data-Structures/Queue/Queue.js [hunk 2]
```diff
@@ -1,56 +1,121 @@
...
+
+  constructor (capacity) {
+    if (capacity < 1) {
+      throw new Error('Capacity is not valid')
```
Comment (appgurueu) [inline Data-Structures/Queue/Queue.js:14]: Why not? Empty queues should be possible; furthermore capacity shouldn't be mandatory.

File: Data-Structures/Queue/Queue.js [hunk 3]
```diff
@@ -1,56 +1,121 @@
...
+
+    this.head = null
+    this.tail = null
+    this.capacity = capacity
+    this.#size = 0
+
+    return Object.seal(this)
+  }
+
+  get length () {
```
Comment (appgurueu) [inline Data-Structures/Queue/Queue.js:25]: I see no need for a getter if `size` were directly exposed.

File: Data-Structures/Queue/Queue.js [hunk 4]
```diff
@@ -1,56 +1,121 @@
...
-  // Add a value to the end of the queue
-  enqueue (item) {
-    this.queue.push(item)
+  /**
+   * @description - Add a value to the end of the queue
+   * @param {*} data
+   * @returns {number} - The current size of queue
+   */
+  enqueue (data) {
+    if (this.capacity === this.#size) {
```
Comment (appgurueu) [inline Data-Structures/Queue/Queue.js:35]: Why should a linked queue have a capacity in the first place?

File: Data-Structures/Queue/Queue.js [hunk 5]
```diff
@@ -1,56 +1,121 @@
...
 
-  // Return Is queue empty ?
-  empty () {
-    return this.queue.length === 0
+  /**
+  * @description - Return is queue empty or not
+  * @returns {boolean}
+  */
+  isEmpty () {
+    return this.length === 0
```
Comment (appgurueu) [inline Data-Structures/Queue/Queue.js:117]: Just use the property and rename it to `length`.

File: Data-Structures/Queue/test/Queue.test.js [hunk 6]
```diff
@@ -13,36 +22,46 @@ describe('Queue', () => {
...
 
-  it('Check empty', () => {
+  it('Check toArray', () => {
     const queue = new Queue()
-    expect(queue.empty()).toBeTruthy()
+
+    queue.enqueue(1)
+    queue.enqueue(2)
+    queue.enqueue(8)
+    queue.enqueue(9)
```
Comment (appgurueu) [inline Data-Structures/Queue/test/Queue.test.js:53]: Missing `expect`, this test does nothing (and you're not passing a capacity?).

**Review by appgurueu — CHANGES_REQUESTED** [review]
Please address the issues and remove all capacity-related code. Otherwise LGTM.

File: Data-Structures/Queue/Queue.js (hunk 1 above)
Comment (fahimfaisaal) [reply Data-Structures/Queue/Queue.js:10]: After constructing the queue, anyone can increase the size of the queue like this `queue.size = 100` for this reason I've used the private prop

Comment (raklaptudirm) [reply Data-Structures/Queue/Queue.js:10]: @appgurueu https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_class_fields

**Review by appgurueu — APPROVED** [review]

**Review by appgurueu — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**